### PR TITLE
Temporal Vectorization Transformation

### PR DESCRIPTION
**Pull request by carljohnsen**
This pull request adds the temporal vectorization transformation, which utilizes the multi-pumping for packing computations temporally on FPGAs.

### CODE REVIEW

File: dace/codegen/CMakeLists.txt [hunk 1]
```diff
@@ -412,6 +415,16 @@ endif() # DACE_ENABLE_RTL
 # Create Xilinx object files
 if(DACE_ENABLE_XILINX)
 
+  if (DACE_XILINX_TARGET_CLOCK MATCHES "|")
+    string(REGEX MATCH "0:([0-9]+)" DACE_XILINX_EXTERNAL_TARGET_CLOCK ${DACE_XILINX_TARGET_CLOCK})
```
Comment (TizianoDeMatteis) [inline dace/codegen/CMakeLists.txt:419]: Apparently many of the CI tests errors are due to these new function (CMAKE errors)

File: dace/codegen/targets/xilinx.py [hunk 2]
```diff
@@ -1002,18 +1056,71 @@ def generate_kernel_internal(self, sdfg: dace.SDFG, state: dace.SDFGState, kerne
...
+                stream_prefix = stream_prefix if multi_pumped else ''
+                kernel_postfix = '_top_1' if multi_pumped else '_1'
+                val = '{}{}.{}{}'.format(kernel_name, kernel_postfix, stream_prefix, stream)
                 self._stream_connections[stream][key] = val
 
         self.generate_modules(sdfg, state, kernel_name, subgraphs, subgraph_parameters, module_stream, entry_stream,
                               state_host_body_stream, instrumentation_stream)
 
+        if multi_pumped:
+            rtllib_config = {
```
Comment (TizianoDeMatteis) [inline dace/codegen/targets/xilinx.py:1073]: pls add comments on why is needed for

File: dace/dtypes.py [hunk 3]
```diff
@@ -74,6 +74,7 @@ class ScheduleType(aenum.AutoNumberEnum):
     FPGA_Device = ()
     Snitch = ()
     Snitch_Multicore = ()
+    FPGA_Double = () #: Used for double pumping
```
Comment (TizianoDeMatteis) [inline dace/dtypes.py:77]: Why just double and not multi-pumped?

**Review by TizianoDeMatteis — COMMENTED** [review]
First round of comments

File: dace/transformation/dataflow/vectorization.py [hunk 4]
```diff
@@ -208,7 +208,7 @@ def apply(self, graph: SDFGState, sdfg: SDFG):
                 else:
                     newlist[contigidx] = (self.vector_len * rb, self.vector_len * rb + self.vector_len - 1, 1)
             edge.data.subset = subsets.Range(newlist)
-            edge.data.volume = vector_size
+            edge.data.volume = 1#vector_size
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/transformation/dataflow/vectorization.py:211]: Ah right, it's a question I have. The StreamingMemory transformation only triggers if volume = 1. However, the Vectorization transformation sets volume = vector_size. But shouldn't it be possible to chain them? 
I'm doing vectorization > fpga > streaming memory > temporal vectorization.

File: dace/codegen/CMakeLists.txt [hunk 5]
```diff
@@ -412,6 +415,16 @@ endif() # DACE_ENABLE_RTL
 # Create Xilinx object files
 if(DACE_ENABLE_XILINX)
 
+  if (DACE_XILINX_TARGET_CLOCK MATCHES "|")
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/codegen/CMakeLists.txt]: The bar is already mentioned in the schema for the xilinx frequency. Would you like it here in the code as well, or is the schema enough?

File: dace/codegen/CMakeLists.txt [hunk 6]
```diff
@@ -430,9 +443,39 @@ if(DACE_ENABLE_XILINX)
...
+
+  # If the project uses generated IP cores (E.g. through double pumping)
+  if(DACE_XILINX_IP_FILES)
+    set(DACE_XILINX_BUILD_FLAGS ${DACE_XILINX_BUILD_FLAGS} --user_ip_repo_paths ip_cores)
+  endif()
+
   unset(DACE_KERNEL_TARGETS)
-  foreach(DACE_KERNEL_FILE ${DACE_XILINX_KERNEL_FILES})
 
+  foreach(DACE_IP ${DACE_XILINX_IP_FILES})
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/codegen/CMakeLists.txt:455]: I've added some comments!

File: dace/dtypes.py (hunk 3 above)
Comment (carljohnsen) [reply dace/dtypes.py]: True! This has been changed

File: dace/transformation/subgraph/temporal_vectorization.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+dp = True
+sched = dtypes.ScheduleType.FPGA_Double if dp else dtypes.ScheduleType.FPGA_Device
+
+@properties.make_properties
+class TemporalVectorization(transformation.SubgraphTransformation):
+    '''
+    This transformation applies the multi-pumping optimization to a subgraph targeting FPGAs, in turn packing more computations temporally rather than spatially as is done in traditional vectorization.
+
+    Currently, it can only be applied to applications targeting Xilinx FPGAs, and to subgraphs that purely communicate through streams. It can be applied in two ways:
+    1. Inwards - where the internal widths of the subgraph are narrowed. This gives the benefit of a reduced critical-resource footprint at the same throughput.
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/transformation/subgraph/temporal_vectorization.py]: Done!

File: dace/codegen/targets/xilinx.py (hunk 2 above)
Comment (carljohnsen) [reply dace/codegen/targets/xilinx.py:1071]: I've added some comments.

File: dace/codegen/CMakeLists.txt (hunk 1 above)
Comment (carljohnsen) [reply dace/codegen/CMakeLists.txt:419]: Yes! This has been fixed, as the regex did not match correctly before.

File: dace/codegen/targets/fpga.py [hunk 8]
```diff
@@ -773,44 +828,66 @@ def make_parameters(self, sdfg: SDFG, state: SDFGState, subgraphs):
...
+            is_external_subgraph = self.is_external_subgraph(subgraph)
+            is_multi_subgraph = self.is_multi_pumped_subgraph(subgraph)
             subsdfg = subgraph.parent
             candidates = []  # type: List[Tuple[bool,str,Data]]
             # [(is an output, dataname string, data object)]
             array_to_banks_used_out: Dict[str, Set[int]] = {}
             array_to_banks_used_in: Dict[str, Set[int]] = {}
-            for n in subgraph.source_nodes():
+            sources = subgraph.source_nodes()
+            # Handle the case where the outermost node in the subgraph is a map
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/codegen/targets/fpga.py]: Yes, it was. However, it does not for the other tests, which is why I have removed it.

File: dace/codegen/targets/fpga.py [hunk 9]
```diff
@@ -563,10 +599,22 @@ def generate_state(self, sdfg: dace.SDFG, state: dace.SDFGState, function_stream
 
                 self._kernels_names_to_id[kernel_name] = kern_id
 
+                if len(multi_sgs) != 0:
+                    self._num_kernels += 1
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/codegen/targets/fpga.py:610]: Currently, there is only added one additional multi pumped per state. In the future, when we can emit multi-pumped kernels that to not consist of directly connected subgraphs, more than 1 should be added to _num_kernels.

File: dace/codegen/targets/fpga.py [hunk 10]
```diff
@@ -388,6 +390,34 @@ def has_initializer(self):
...
+            if isinstance(n, dace.nodes.NestedSDFG):
+                for sg in dace.sdfg.concurrent_subgraphs(n.sdfg.start_state):
+                    if self.is_external_subgraph(sg):
+                        return True
+            elif any([
+                isinstance(n, dace.nodes.Tasklet) and n.language == dace.dtypes.Language.SystemVerilog]):
+                return True
+        return False
+    
+    def is_multi_pumped_subgraph(self, subgraph):
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/codegen/targets/fpga.py]: This has been added!

File: dace/codegen/targets/fpga.py [hunk 11]
```diff
@@ -388,6 +390,34 @@ def has_initializer(self):
     def has_finalizer(self):
         return False
 
+    def is_rtl_tasklet(self, subgraph):
+        for n in subgraph.nodes():
+            if (isinstance(n, dace.nodes.Tasklet) and n.language == dace.dtypes.Language.SystemVerilog):
+                return n
+        return None
+
+    def is_external_subgraph(self, subgraph):
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/codegen/targets/fpga.py]: This has been added!

File: dace/codegen/targets/fpga.py [hunk 12]
```diff
@@ -388,6 +390,34 @@ def has_initializer(self):
     def has_finalizer(self):
         return False
 
+    def is_rtl_tasklet(self, subgraph):
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/codegen/targets/fpga.py]: This has been added!

File: dace/codegen/CMakeLists.txt [hunk 13]
```diff
@@ -430,9 +443,39 @@ if(DACE_ENABLE_XILINX)
     set(DACE_XILINX_MINIMUM_FIFO_DEPTH "\nconfig_dataflow -fifo_depth ${DACE_MINIMUM_FIFO_DEPTH}")
   endif()
 
+
+  # If the project uses generated IP cores (E.g. through double pumping)
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/codegen/CMakeLists.txt]: Fixed!

File: dace/codegen/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (TizianoDeMatteis) [reply dace/codegen/CMakeLists.txt]: Nope, in the schema is enough!

File: dace/transformation/dataflow/vectorization.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TizianoDeMatteis) [reply dace/transformation/dataflow/vectorization.py:211]: Yes. We should fix streaming memory then.
@tbennun any thoughts on this?

[the remark this answers is not in the document]
Comment (tbennun) [reply dace/transformation/dataflow/vectorization.py:211]: Because the edge.data.data dtype is, e.g., float64, this should stay `vector_size`. StreamingMemory should change accordingly though.

File: dace/codegen/targets/fpga.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (TizianoDeMatteis) [reply dace/codegen/targets/fpga.py:610]: Ok, pls add a comment about this

File: dace/transformation/dataflow/vectorization.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/transformation/dataflow/vectorization.py]: I have removed this hack, and adjusted the check inside StreamingMemory instead. Now the hardware tests work as expected.

File: dace/codegen/targets/fpga.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (carljohnsen) [reply dace/codegen/targets/fpga.py:613]: This has been added!

File: dace/codegen/targets/fpga.py [hunk 14]
```diff
@@ -388,6 +392,39 @@ def has_initializer(self):
     def has_finalizer(self):
         return False
 
+    def is_rtl_subgraph(self, subgraph: ScopeSubgraphView):
+        '''
+        Checks whether the given subgraph is an RTL subgraph, as in: does it contain a SystemVerilog tasklet.
```
Comment (TizianoDeMatteis) [inline dace/codegen/targets/fpga.py:397]: Question mark?

File: dace/codegen/targets/fpga.py [hunk 15]
```diff
@@ -388,6 +392,39 @@ def has_initializer(self):
...
+                    node = self.is_rtl_subgraph(sg)
+                    if node:
+                        return node
+            elif isinstance(n, dace.nodes.Tasklet) and n.language == dace.dtypes.Language.SystemVerilog:
+                return n
+        return None
+
+    def is_multi_pumped_subgraph(self, subgraph: ScopeSubgraphView):
+        '''
+        Checks whether the given subgraph is a multi-pumped subgraph, as in: does it contain a map whose schedule is set to multi-pumped.
```
Comment (TizianoDeMatteis) [inline dace/codegen/targets/fpga.py:414]: Question mark?

File: dace/transformation/subgraph/temporal_vectorization.py [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        'Which approach to use. Can be 1 (increased throughput, same resources) or 2 (same throughput, reduced resourced).'
+    )
+
+    def can_be_applied(self, sdfg: SDFG, subgraph: SubgraphView) -> bool:
+        '''
+        Temporal vectorization can be applied if:
+        1. There is one outermost map in the subgraph. 
+        2. All of the non- source and sink nodes are only accessed within this subgraph. 
+        3. All of the source and sink nodes are either streams or scalars. 
+        4. If the direction is 1, then either:
```
Comment (TizianoDeMatteis) [inline dace/transformation/subgraph/temporal_vectorization.py:38]: What do you mean by "the direction is 1"?

Comment (TizianoDeMatteis) [reply dace/transformation/subgraph/temporal_vectorization.py:38]: Ah, I think you are referring to the two "ways" defined before. Pls make it more clear. Later on you call this approach

File: tests/rtl/hardware_test.py [hunk 17]
```diff
@@ -385,6 +388,73 @@ def test_hardware_axpy_double_pump_vec4():
     return test_hardware_axpy_double_pump(veclen=4)
 
 
+@rtl_test()
+def test_hardware_vadd_temporal_vectorization():
+    # TODO !!!!! THIS TEST STALLS WITH VITIS 2021.2 !!!!!
```
Comment (TizianoDeMatteis) [inline tests/rtl/hardware_test.py:393]: oh dear lord

File: tests/rtl/hardware_test.py [hunk 18]
```diff
@@ -385,6 +388,73 @@ def test_hardware_axpy_double_pump_vec4():
...
+        result = np.zeros(N.get(), dtype=np.float32)
+        expected = x + y
+
+        # Generate the initial SDFG
+        def np_vadd(x: dace.float32[N], y: dace.float32[N]):
+            return x + y
+
+        sdfg = dace.program(np_vadd).to_sdfg()
+
+        # Remove underscores as Xilinx does not like them
```
Comment (TizianoDeMatteis) [inline tests/rtl/hardware_test.py:412]: Why is this needed here in a test?
Should the user do the same in her own program?

File: tests/rtl/hardware_test.py [hunk 19]
```diff
@@ -385,6 +388,73 @@ def test_hardware_axpy_double_pump_vec4():
     return test_hardware_axpy_double_pump(veclen=4)
 
 
+@rtl_test()
+def test_hardware_vadd_temporal_vectorization():
+    # TODO !!!!! THIS TEST STALLS WITH VITIS 2021.2 !!!!!
+    # But it works fine for 2020.2
```
Comment (TizianoDeMatteis) [inline tests/rtl/hardware_test.py:394]: Could you add small description of the test?

**Comment by tbennun** [comment]
@carljohnsen I approve the changes made to the transformation.

File: dace/transformation/subgraph/temporal_vectorization.py (hunk 16 above)
Comment (carljohnsen) [reply dace/transformation/subgraph/temporal_vectorization.py]: Right! I missed that one. I changed from 'direction' to 'approach'.

File: tests/rtl/hardware_test.py (hunk 17 above)
Comment (carljohnsen) [reply tests/rtl/hardware_test.py:431]: ( ͡° ͜ʖ ͡°)

File: tests/rtl/hardware_test.py (hunk 18 above)
Comment (carljohnsen) [reply tests/rtl/hardware_test.py]: This has been fixed, so it's now handled in the Xilinx codegen. So no, it should not have to be done by the user.

File: tests/rtl/hardware_test.py (hunk 19 above)
Comment (carljohnsen) [reply tests/rtl/hardware_test.py:432]: I've added a description to this test, and the others in `hardware_test.py`.

File: tests/rtl/hardware_test.py (hunk 17 above)
Comment (TizianoDeMatteis) [reply tests/rtl/hardware_test.py:431]: Do you have an idea why?
Maybe you can a short comment on this

Comment (carljohnsen) [reply tests/rtl/hardware_test.py:431]: It seems like it's the last transaction somewhere inside that messes up, as it receives all of it's input, but the output is missing a single transaction. It looked like it was the floating point adder was the culprit in hardware emulation. I would have to double check though, in order to be certain to this claim.

**Review by TizianoDeMatteis — APPROVED** [review]
